Binance/BinanceUS: Fix various live test failures

gbjk opened the pull request:
* Fixes failures on `TestGetAggregatedTradesBatched` and `TestGetHistoricTrades`
* Refactors and simplifies `GetAggregatedTradesBatched` (hopefully)

Note: These fixes are for testing with `--tags mock_test_off`

Fixes #2010

- [x] Test fix (non-breaking change which fixes an issue)

shazbert commented on exchanges/binance/binance.go:
```diff
@@ -311,37 +309,33 @@ func (e *Exchange) batchAggregateTrades(ctx context.Context, arg *AggregatedTrad
-			})
+		switch len(additionalTrades) {
+		case 0, 1:
+			break // We only got the one we already have
```
break outer loop? or just rm to hit if case below?

shazbert commented on exchanges/binance/binance_test.go:
```diff
@@ -1477,109 +1477,108 @@ func TestNewOrderTest(t *testing.T) {
 func TestGetHistoricTrades(t *testing.T) {
```
This fails on a live test

shazbert commented on exchanges/binance/binance_test.go:
```diff
@@ -1477,109 +1477,108 @@ func TestNewOrderTest(t *testing.T) {
+// TestGetAggregatedTradesBatched exercises TestGetAggregatedTradesBatched to ensure our date and limit scanning works correctly
+// This test is susceptible to failure if volumes change a lot, during wash trading or zero-fee periods
+// In live tests, 45 minutes is expected to return more than 1000 records
 func TestGetAggregatedTradesBatched(t *testing.T) {
```
this fails on a live test

gbjk replied:
Thanks!
Could you be more specific?
Did it fail locally for you, and can you replicate now?

```
❯ go test -count=1 --tags mock_test_off -v  ./exchanges/binance/ -run TestGetAggregatedTradesBatched/
--- PASS: TestGetAggregatedTradesBatched (0.00s)
    --- PASS: TestGetAggregatedTradesBatched/custom_limit_with_start_time_set,_no_end_time (0.46s)
    --- PASS: TestGetAggregatedTradesBatched/batch_with_timerange (4.77s)
PASS
ok      github.com/thrasher-corp/gocryptotrader/exchanges/binance       8.984s
```

shazbert commented on exchanges/binance/binance.go:
```diff
@@ -268,40 +268,38 @@ func (e *Exchange) GetAggregatedTrades(ctx context.Context, arg *AggregatedTrade
-func (e *Exchange) batchAggregateTrades(ctx context.Context, arg *AggregatedTradeRequestParams, params url.Values) ([]AggregatedTrade, error) {
+// If no args.FromID is passed, requests are made intervals doubling from 10s until we get a viable starting position.
+// Once the starting set is found, we continue scanning until we have all the trades in the time window
+func (e *Exchange) batchAggregateTrades(ctx context.Context, args *AggregatedTradeRequestParams, params url.Values) ([]AggregatedTrade, error) {
```
Seems a bit backwards. I think this functionality should be in the wrapper file and only called in `GetHistoricTrades` , and GetAggregatedTrades get called without the batch handling and the added complexity and return exactly what it needs. Just to map to API. But not worth changing.

shazbert reviewed (changes requested):
Just some nits for live testing ❤️

thrasher- commented:
@codex please review

thrasher- reviewed (changes requested):
Looks like BinanceUS tests are failing:

https://github.com/thrasher-corp/gocryptotrader/actions/runs/17791729939/job/50570112052#step:11:3081

gloriousCode commented on exchanges/binance/binance.go:
```diff
@@ -268,40 +268,42 @@ func (e *Exchange) GetAggregatedTrades(ctx context.Context, arg *AggregatedTrade
 	}
```
[You thanked me too soon ](https://github.com/thrasher-corp/gocryptotrader/pull/2019#pullrequestreview-3236712285) (sorry)

gbjk commented:
> Looks like BinanceUS tests are failing:

@thrasher- Thanks for spotting that.
We lost the symbol formatting in the changes to binanceus.
I've fixed that, and also re-read the patch more carefully and verified the 59 minute clipping is not necessary.

gloriousCode commented on exchanges/binanceus/binanceus_test.go:
```diff
@@ -1756,3 +1760,50 @@ func TestGetCurrencyTradeURL(t *testing.T) {
+				t.Helper()
+				// 2000 records in was about 6 minutes in 2025; Adjust if Binance enters a phase of zero-fees or low-volume
+				require.Equal(t, 2042, len(results), "must return exactly the limit number of records")
+				assert.WithinDuration(t, results[len(results)-1].TimeStamp.Time(), start, 48*time.Hour, "last record should be within 20 minutes of start time")
```
```suggestion
				assert.WithinDuration(t, results[len(results)-1].TimeStamp.Time(), start, 48*time.Hour, "last record should be within 48 hours of start time")
```
binanceus is the place for trading solo

gbjk replied:
Reviewed all of the numbers with live results and updated accordingly.

Fixed https://github.com/gbjk/gocryptotrader/commit/7fd911d1b

If this turns out to be fragile can just widen the hole ....

gloriousCode reviewed (changes requested):
minor things. thanks for applying (and also fixing) my diff!

gbjk commented on exchanges/binanceus/binanceus.go:
```diff
@@ -203,84 +202,81 @@ func (e *Exchange) GetHistoricalTrades(ctx context.Context, hist HistoricalTrade
 		return nil, err
+	} else {
+		params.Set("symbol", s)
 	}
```
Sorry, I just spotted that this is a lint revive error.
How hateful 😡 

Fixed https://github.com/gbjk/gocryptotrader/commit/bef5158bf

gloriousCode reviewed (approved):
Thanks for making those changes!

gbjk commented:
> Thanks for resolving this issue! Live tests are happy for candle stuff but there's one last live test which isn't happy, so please apply this patch:

Fixed https://github.com/gbjk/gocryptotrader/commit/185def555

> There's also a linter issue after rebasing

Fixed https://github.com/gbjk/gocryptotrader/commit/2c70d16f2

thrasher- reviewed (approved):
Changes tACK, thanks!

shazbert reviewed (approved):
Thanks!
